### ISSUE TITLE
Ltest-mem-validate and other test case fail on linux

### ISSUE THREAD

**Issue by ziyangc97** [libunwind/libunwind#489]
hi, I am runing make check of libunwind 1.6.2 on openEuler(linux) arm system and there are 4 test cases failed.
[test-suite.log](https://github.com/libunwind/libunwind/files/11239083/test-suite.log)

My OS and libunwind info:
kernel:5.10.0-60.18.0.50.oe2203.aarch64
aarch64
libunwind:1.6.2
libtool:libtool-2.4.6-34

I try to debug but found out and stuck here:
file Ltest-mem-validate.c
```
if (childpid)
        {
          wait(&status);
          printf("iter i = %d, WIFEXITED(status) = %d\n", i, WIFEXITED(status));
          if (WIFEXITED(status))
              return WEXITSTATUS(status);
          else {
            printf("return -1\n");
```
the test-suite.log shows:
```
iter i = 3, WIFEXITED(status) = 0
return -1
FAIL Ltest-mem-validate (exit status: 255)
```

Please provide some hint why this is happening or some guidance how to singly run this test because I try to run it by gdb.

Appreciate anyone's help

**Comment by bregma** [comment]
At the very least the return value of `wait()` needs to be checked for success or failure. As to why it would be failing (likely ECHILD, but possibly EINTR) will take more investigation.

**Comment by bregma** [comment]
I suspect the real reason for failure is that the stack frame for 5 or more recursive calls of `consume_and_run()` may exceed 4k on aarch64. If you have the time, that's where I'd start investigating.

**Comment by bregma** [comment]
Looking further, I think this test is bogus and highly reliant on (a) undefined behaviour (strict aliasing violation) and (2) ABI-specific behaviour (eg. stack growth direction, calling conventions). That it works on x86* is a fluke.

The test code contains the comment
>     We make the assumption that we are able to rewind far enough
>     (steps > 5) before touching the forbidden region in the stack,
>     at which point the unwinding should stop gracefully.
but if the unwinder hits the mprotected region (which it will, at least on ARM ABIs), the OS is supposed to raise a segfaukt, which the unwinder does not handle gracefully.

It's not clear what this test is attempting to validate (other than that the unwinder detects artificial restrictions on memory access, which it does not do and of course the test usually fails). If anyone could enlighten me on the purpose of this test I'm all ears otherwise I think it should just be removed.

**Comment by he7850** [comment]
In my view, Ltest-mem-validate testcase intent is:

1. In the child process, create a 10-layer call stack, set the memory page which the initial stack start address belongs to as unreadable, and then use 'unw_step' function of libunwind to traverse and access each layer of the call stack, and print the ip and sp address
2. It is expected that in the child process, it can traverse at least 5 layers of the call stack, and when it touches the stack area that is forbidden for access, 'unw_step' function can exit gracefully by mem_validate()

**Comment by he7850** [comment]
> Looking further, I think this test is bogus and highly reliant on (a) undefined behaviour (strict aliasing violation) and (2) ABI-specific behaviour (eg. stack growth direction, calling conventions). That it works on x86* is a fluke.
> 
> The test code contains the comment
> 
> > ```
> > We make the assumption that we are able to rewind far enough
> > (steps > 5) before touching the forbidden region in the stack,
> > at which point the unwinding should stop gracefully.
> > ```
> 
> but if the unwinder hits the mprotected region (which it will, at least on ARM ABIs), the OS is supposed to raise a segfaukt, which the unwinder does not handle gracefully.
> 
> It's not clear what this test is attempting to validate (other than that the unwinder detects artificial restrictions on memory access, which it does not do and of course the test usually fails). If anyone could enlighten me on the purpose of this test I'm all ears otherwise I think it should just be removed.

The reason why child process returns -1: the unw_step function triggers a segment error (Segmentation fault)
gdb trace:
```
(gdb) n
73	      ret = unw_step (&cursor);
(gdb) n
Thread 2.1 "lt-Ltest-mem-va" received signal SIGSEGV, Segmentation fault.
access_mem (as=<optimized out>, addr=281474976699672, val=0xffffffff0be0, write=<optimized out>, arg=<optimized out>) at aarch64/Ginit.c:337
337	      *val = *(unw_word_t *) addr;
(gdb) bt
#0  access_mem (as=<optimized out>, addr=281474976699672, val=0xffffffff0be0, write=<optimized out>, arg=<optimized out>) at aarch64/Ginit.c:337
#1  0x0000fffff7f75018 in dwarf_get (val=<optimized out>, loc=..., c=<optimized out>) at ../include/tdep-aarch64/libunwind_i.h:246
#2  apply_reg_state (c=c@entry=0xffffffff1e98, rs=rs@entry=0xffffffff1670) at dwarf/Gparser.c:885
#3  0x0000fffff7f76bf0 in _Uaarch64_dwarf_step (c=c@entry=0xffffffff1e98) at dwarf/Gparser.c:974
#4  0x0000fffff7f719b0 in _Uaarch64_step (cursor=0xffffffff1e98) at aarch64/Gstep.c:162
#5  0x0000aaaaaaaa0ea0 in do_backtrace () at Ltest-mem-validate.c:73
#6  0x0000aaaaaaaa0fd0 in consume_and_run (depth=0) at Ltest-mem-validate.c:102
```

**Comment by bregma** [comment]
Fixed by #494
